layers: Move ValidationStateTracker into state_tracker.h & .cpp

locke-lunarg opened the pull request:
Some static functions that were used in both of `ValidationStateTracker & CoreChecks` were moved into `state_tracker.h`.

mark-lunarg commented on layers/CMakeLists.txt:
```diff
@@ -162,6 +162,7 @@ set(CHASSIS_LIBRARY_FILES
     generated/layer_chassis_dispatch.cpp)
 
 set(CORE_VALIDATION_LIBRARY_FILES
+    state_tracker.cpp
```
This file is or will be used by BestPractices and GpuValidation.  Should it get moved to CHASSIS_LIBRARY_FILES, or duplicated in those two groups as well?

mark-lunarg replied:
John votes for CHASSIS_LIBRARY_FILES instead of CORE_VALIDATION_LIBRARY_FILES.

mark-lunarg commented on scripts/vk_validation_stats.py:
```diff
@@ -51,6 +51,7 @@
 
 layer_source_files = [common_codegen.repo_relative(path) for path in [
     'layers/buffer_validation.cpp',
+    'layers/state_tracker.cpp',
```
I don't think we need this here -- it doesn't (and should never) have any VUIDs in it, as with the chassis files.

jzulauf-lunarg commented on layers/state_tracker.cpp:
```diff
@@ -0,0 +1,4595 @@
+                    pSemaphore->signaled = false;
+                } else {
+                    semaphore_externals.push_back(semaphore);
+                    pSemaphore->in_use
```
Leave this and the other shader_module utilites in shader_validation.cpp as non-static functions.

jzulauf-lunarg commented on layers/state_tracker.cpp:
```diff
@@ -0,0 +1,4595 @@
+#include "vk_typemap_helper.h"
+
+#include "chassis.h"
+#include "core_validation.h"
```
state_tracker.cpp shouldn't include core_validation.h --  if it does there is something to clean up.

locke-lunarg commented on layers/shader_validation.cpp:
```diff
@@ -232,16 +226,6 @@ unsigned ExecutionModelToShaderStageFlagBits(unsigned mode) {
     }
 }
 
-static spirv_inst_iter FindEntrypoint(SHADER_MODULE_STATE const *src, char const *name, VkShaderStageFlagBits stageBits) {
```
`FindEntrypoint` is used in `ValidationStateTracker::RecordPipelineShaderStage & CoreChecks::ValidateGraphicsPipelineShaderState & CoreChecks::ValidateComputePipeline &  CoreChecks::ValidateRayTracingPipelineNV`, so it's moved into `state_tracker.h`.

locke-lunarg commented on layers/shader_validation.cpp:
```diff
@@ -561,31 +545,6 @@ static unsigned GetFormatType(VkFormat fmt) {
 
-// characterizes a SPIR-V type appearing in an interface to a FF stage, for comparison to a VkFormat's characterization above.
-// also used for input attachments, as we statically know their format.
-static unsigned GetFundamentalType(SHADER_MODULE_STATE const *src, unsigned type) {
```
`GetFundamentalType `is used in `DescriptorTypeToReqs`, and then `DescriptorTypeToReqs `is used in `ValidationStateTracker::RecordPipelineShaderStage`, , so it's moved into `state_tracker.h`.

locke-lunarg commented on layers/shader_validation.cpp:
```diff
@@ -847,75 +806,6 @@ static std::vector<std::pair<uint32_t, interface_var>> CollectInterfaceByInputAt
     return out;
 }
 
-static bool IsWritableDescriptorType(SHADER_MODULE_STATE const *module, uint32_t type_id, bool is_storage_buffer) {
```
`IsWritableDescriptorType`is used in `CollectInterfaceByDescriptorSlot`, and then `CollectInterfaceByDescriptorSlot`is used in `ValidationStateTracker::RecordPipelineShaderStage`, so it's moved into `state_tracker.cpp`.

locke-lunarg commented:
Moved `shader_module` functions code back to `shader_validation.cpp` and removed "static" functions in `state_tracker.cpp & shader_validation.cpp`.

jzulauf-lunarg commented:
Not moved from core_validation.cpp

`ValidationStateTracker::ResetCommandBufferPushConstantDataIfIncompatible`

jzulauf-lunarg commented on layers/state_tracker.h:
```diff
@@ -0,0 +1,980 @@
+#include "core_validation_error_enums.h"
+#include "core_validation_types.h"
+#include "descriptor_sets.h"
+#include "shader_validation.h"
```
are shader_validation and gpu validation headers needed?  can we forward declare what's needed?

locke-lunarg commented:
@jzulauf-lunarg 

Made declarations of some shader common functions in `shader_validation.h` and keep them in `shader_validation`.

Moved the whole `ValidationStateTracker` into `state_tracker`.

Moved  `#include "shader_validation.h"` from `.h` to `.cpp` in `state_tracker`.
